pg_replicate/src/clients/delta.rs

```diff
@@ -0,0 +1,235 @@
+use deltalake::operations::create::CreateBuilder;
+use deltalake::{kernel::DataType, DeltaOps, DeltaTableError};
+use std::{collections::HashMap, path::Path, sync::Arc};
+use tokio_postgres::types::Type;
+
+use crate::{
+    conversions::{table_row::TableRow, Cell},
+    table::{ColumnSchema, TableId, TableName, TableSchema},
+};
+use deltalake::arrow::array::{
+    Array, BooleanArray, Date32Array, Float32Array, Float64Array, Int32Array,
+    RecordBatch as DeltaRecordBatch, StringArray, TimestampMicrosecondArray, UInt32Array,
+};
+
+pub struct DeltaClient {
+    pub path: String,
+    pub table_schemas: Option<HashMap<TableId, TableSchema>>,
+    pub delta_schemas: Option<HashMap<String, Arc<Schema>>>,
+}
+
+impl Default for DeltaClient {
+    fn default() -> Self {
+        Self {
+            path: Default::default(),
+            table_schemas: Default::default(),
+            delta_schemas: Default::default(),
+        }
+    }
+}
+impl DeltaClient {
+    fn naive_date_to_arrow(date: NaiveDate) -> i32 {
+        let epoch = NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
+        (date - epoch).num_days() as i32
+    }
+
+    fn postgres_to_delta(typ: &Type) -> DataType {
+        match typ {
+            &Type::BOOL => DataType::BOOLEAN,
+            &Type::CHAR | &Type::BPCHAR | &Type::VARCHAR | &Type::NAME | &Type::TEXT => {
+                DataType::STRING
+            }
+            &Type::INT2 | &Type::INT4 | &Type::INT8 => DataType::INTEGER,
+            &Type::FLOAT4 | &Type::FLOAT8 | &Type::NUMERIC => DataType::FLOAT,
+            &Type::DATE => DataType::DATE,
+            &Type::TIME | &Type::TIMESTAMP | &Type::TIMESTAMPTZ => DataType::TIMESTAMP,
+            &Type::UUID => DataType::STRING,
+            &Type::OID => DataType::INTEGER,
+            &Type::BYTEA => DataType::BYTE,
+            _ => DataType::STRING,
+        }
+    }
+
+    fn postgres_to_arrow(typ: &Type) -> ArrowDataType {
+        match typ {
+            &Type::BOOL => ArrowDataType::Boolean,
+            &Type::CHAR | &Type::BPCHAR | &Type::VARCHAR | &Type::NAME | &Type::TEXT => {
+                ArrowDataType::Utf8
+            }
+            &Type::INT2 | &Type::INT4 | &Type::INT8 => ArrowDataType::Int32,
+            &Type::FLOAT4 | &Type::FLOAT8 | &Type::NUMERIC => ArrowDataType::Float16,
+            &Type::DATE => ArrowDataType::Date32,
+            &Type::TIME | &Type::TIMESTAMP | &Type::TIMESTAMPTZ => ArrowDataType::Utf8,
+            &Type::UUID => ArrowDataType::Utf8,
+            &Type::OID => ArrowDataType::Utf8,
+            _ => ArrowDataType::Utf8,
+        }
+    }
+
+    pub async fn create_table_schema(
+        &mut self,
+        tb_name: String,
+        columns: &[ColumnSchema],
+    ) -> Result<Arc<Schema>, DeltaTableError> {
+        let full_path = self.delta_full_path(tb_name.clone());
+
+        if Path::new(&full_path).join("_delta_log").exists() {
+            DeltaOps::try_from_uri(&full_path).await?.delete().await?;
```

Comment:
Why delete if the path exists? Shouldn't it continue to use the path if it exists.

Reply:
yesss